### PR TITLE
Lyrics in game

### PR DESCRIPTION
**Pull request by sim0mo**
If applied, this PR will add the final functionality concerning lyrics : they will be displayed in game, only to the player that is currently singing. He can chose to scroll them to sing along, or simply close the popup window. If it does, it can still resume this window by clicking on a button.

### CODE REVIEW

File: app/src/main/java/com/github/fribourgsdp/radio/GameActivity.kt [hunk 1]
```diff
@@ -184,4 +189,42 @@ open class GameActivity : AppCompatActivity(), GameView {
...
+//
+////        // dismiss the popup window when touched
+//        popupView.findViewById<ImageView>(R.id.close_popup_button).setOnClickListener{
+//            lyricsPopup.dismiss()
+//        }
+//        popupView.setOnTouchListener { v, _ ->
+//            v.performClick()
+//            lyricsPopup.dismiss()
+//            true
+//        }
```
Comment (shanzhaiii) [inline app/src/main/java/com/github/fribourgsdp/radio/GameActivity.kt:228]: is there a reason for keeping this in the PR? I think since we have versioning, this could be removed

**Review by shanzhaiii — APPROVED** [review]
Not much to say appart from the commented out code which I think should be removed. Other than that, feature seems well implemented and I like that you are using a popup fragment ;)

File: app/src/main/java/com/github/fribourgsdp/radio/GameActivity.kt (hunk 1 above)
Comment (sim0mo) [reply app/src/main/java/com/github/fribourgsdp/radio/GameActivity.kt:228]: Oh you're right, I can be oblivious sometimes when it comes to comments and debug messages, I shall remove it.

File: app/src/main/java/com/github/fribourgsdp/radio/MusixmatchLyricsGetter.kt [hunk 2]
```diff
@@ -124,10 +127,14 @@ object MusixmatchLyricsGetter : LyricsGetter {
...
-                    ?.getJSONObject("message")
-                    ?.getJSONObject("body")
-                    ?.getJSONArray("track_list")
+                val trackList = try {
+                    parsedResponse
+                        ?.getJSONObject("message")
+                        ?.getJSONObject("body")
+                        ?.getJSONArray("track_list")
+                } catch (e : Exception){
+                    JSONArray()
```
Comment (nathanduchesne) [inline app/src/main/java/com/github/fribourgsdp/radio/MusixmatchLyricsGetter.kt:136]: As discussed IRL, we can try and find the source of the error causing this exception to be thrown next week.

File: app/src/main/java/com/github/fribourgsdp/radio/SnapshotUtility.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package com.github.fribourgsdp.radio
+
+import com.google.firebase.firestore.DocumentSnapshot
```
Comment (nathanduchesne) [inline app/src/main/java/com/github/fribourgsdp/radio/SnapshotUtility.kt:3]: Thank you for implementing this class, which makes the code reading a bit clearer when accessing the database snapshots.

**Review by nathanduchesne — APPROVED** [review]

File: app/src/main/java/com/github/fribourgsdp/radio/MusixmatchLyricsGetter.kt (hunk 2 above)
Comment (sim0mo) [reply app/src/main/java/com/github/fribourgsdp/radio/MusixmatchLyricsGetter.kt:136]: Yes that was an emergency patch.

File: app/src/main/java/com/github/fribourgsdp/radio/SnapshotUtility.kt (hunk 3 above)
Comment (sim0mo) [reply app/src/main/java/com/github/fribourgsdp/radio/SnapshotUtility.kt:3]: You're welcome

File: app/src/androidTest/java/com/github/fribourgsdp/radio/PlayerGameHandlerTest.kt [hunk 4]
```diff
@@ -33,6 +34,15 @@ class PlayerGameHandlerTest {
...
         `when`(mockSnapshot.getBoolean("finished")).thenReturn(false)
+        mockSnapshot2 = mock(DocumentSnapshot::class.java)
+        `when`(mockSnapshot2.getString("singer")).thenReturn("")
+        `when`(mockSnapshot2.exists()).thenReturn(true)
+        `when`(mockSnapshot2.getLong("current_round")).thenReturn(round)
+        `when`(mockSnapshot2.get("song_choices")).thenReturn(listOfSongs)
+        `when`(mockSnapshot2.get("song_choices_lyrics")).thenReturn(hashMapOf("Momentum" to "Lorem Ipsum"))
+        `when`(mockSnapshot2.getString("current_song")).thenReturn("Momentum")
+        `when`(mockSnapshot2.get("scores")).thenReturn(scores)
+        `when`(mockSnapshot2.getBoolean("finished")).thenReturn(false)
```
Comment (VictorSchneuwly) [inline app/src/androidTest/java/com/github/fribourgsdp/radio/PlayerGameHandlerTest.kt:45]: Since you only use `mocksnapshot2` once you could simply write this on `mocksnapshot` before using this in your test. That way this isn't computed before every test and only used once.

**Review by VictorSchneuwly — CHANGES_REQUESTED** [review]
Extending the snapshot is a good idea for clarity but next time you want to change something like this it should be in another pr as this will most likely result in conflicts.

File: app/src/main/res/layout/activity_game.xml [hunk 5]
```diff
@@ -68,13 +68,32 @@
             android:visibility="gone"
             android:layout_marginTop="24dp" />
 
-        <Button
-            android:id="@+id/songGuessSubmitButton"
-            android:layout_width="wrap_content"
-            android:layout_height="wrap_content"
+        <LinearLayout
```
Comment (VictorSchneuwly) [inline app/src/main/res/layout/activity_game.xml:71]: Why did you add another linear layout ?

File: app/src/main/res/layout/activity_game.xml [hunk 6]
```diff
@@ -68,13 +68,32 @@
...
+                android:textSize="16sp"
+                style="?android:attr/buttonBarButtonStyle" />
+
+            <Button
+                android:id="@+id/showLyricsButton"
+                style="?android:attr/buttonBarButtonStyle"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:textSize="16sp"
+                android:text="@string/show_lyrics" />
```
Comment (VictorSchneuwly) [inline app/src/main/res/layout/activity_game.xml:91]: You should init its visibility to `gone` because otherwise, the first guesser sees this button even when he souldn't.

**Review by VictorSchneuwly — CHANGES_REQUESTED** [review]
After some tests on my emulator, I found those two UI remarks.

Also, while playing I encountered a bug where, after closing the lyrics display, I could still pick a song even when all other players guessed. This resulted in the game crashing. Maybe you should add a `hideLyrics` method to `GameView` that would close the popup when it's not your turn to sing anymore.

File: app/src/main/res/layout/activity_game.xml (hunk 6 above)
Comment (sim0mo) [reply app/src/main/res/layout/activity_game.xml:91]: Ok !

File: app/src/main/res/layout/activity_game.xml (hunk 5 above)
Comment (VictorSchneuwly) [reply app/src/main/res/layout/activity_game.xml:71]: We can remove it later as those two buttons will never be shown at the same time.

**Comment by sim0mo** [comment]
> After some tests on my emulator, I found those two UI remarks.
> 
> Also, while playing I encountered a bug where, after closing the lyrics display, I could still pick a song even when all other players guessed. This resulted in the game crashing. Maybe you should add a `hideLyrics` method to `GameView` that would close the popup when it's not your turn to sing anymore.

Indeed, I had not considered this scenario... This will be a bug fix task for me next sprint.

**Review by VictorSchneuwly — APPROVED** [review]
It's all good for me now!
